File: python/prototype/prototype.py
from abc import ABC


class Prototype(ABC):
    def clone(self) -> "Prototype":
        raise NotImplementedError()


class Shape(Prototype):
    def __init__(self, x: int, y: int):
        self._x = x
        self._y = y

    def clone(self) -> "Prototype":
        return Shape(self._x, self._y)

    def show(self):
        print(self._x, self._y)


def main():
    shape1 = Shape(2, 3)
    shape2: Shape = shape1.clone()
    shape1.show()
    shape2.show()


if __name__ == '__main__':
    main()

File: cpp/prototype/prototype.cpp
#include <iostream>
#include <memory>


class Prototype {
public:
    virtual std::shared_ptr<Prototype> clone() = 0;
};

class Shape : public Prototype {
public:
    Shape(int x, int y) : _x(x), _y(y) {
    }

    void show() {
        std::cout << _x << " " << _y << std::endl;
    }

    std::shared_ptr<Prototype> clone() override {
        return std::make_shared<Shape>(_x, _y);
    }

private:
    int _x;
    int _y;
};


int main() {
    Shape shape1(2, 31);
    auto shape2 = shape1.clone();
    shape1.show();
    reinterpret_cast<Shape *>(shape2.get())->show();
    return 0;
}
